applications/EigenSolversApplication/custom_solvers/eigen_generalized_eigenvalue_solver.h

```diff
@@ -0,0 +1,185 @@
+/*
+//  KRATOS _______
+//        / ____(_)___ ____  ____
+//       / __/ / / __ `/ _ \/ __ \
+//      / /___/ / /_/ /  __/ / / /
+//     /_____/_/\__, /\___/_/ /_/ SolversApplication
+//             /____/
+//
+//  Author: Armin Geiser
+*/
+
+#if !defined(EIGEN_GENERALIZED_EIGEN_SOLVER_H_INCLUDED)
+#define EIGEN_GENERALIZED_EIGEN_SOLVER_H_INCLUDED
+
+// External includes
+
+#include <Eigen/Core>
+#include <Eigen/Eigenvalues>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "linear_solvers/linear_solver.h"
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+/// This solver solves the generalized eigenvalue problem using Eigen.
+/// it solves for all eigenvalues
+template<class TSparseSpaceType, class TDenseSpaceType,
+        class TReordererType = Reorderer<TSparseSpaceType, TDenseSpaceType> >
+class EigenGeneralizedEigenvalueSolver: public LinearSolver<TSparseSpaceType, TDenseSpaceType,
+        TReordererType> {
+
+  public:
+    KRATOS_CLASS_POINTER_DEFINITION(EigenGeneralizedEigenvalueSolver);
+
+    typedef LinearSolver<TSparseSpaceType, TDenseSpaceType, TReordererType> BaseType;
+
+    typedef typename TSparseSpaceType::MatrixType SparseMatrixType;
+
+    typedef typename TSparseSpaceType::VectorType DenseVectorType;
+
+    typedef typename TDenseSpaceType::MatrixType DenseMatrixType;
+
+    typedef std::size_t SizeType;
+
+    EigenGeneralizedEigenvalueSolver(Parameters param) : mParam(param)
+    {
+
+        Parameters default_params(R"(
+        {
+            "solver_type": "generalized_eigenvalue_solver",
+            "echo_level": 1
+        })");
+
+        // don't validate linear_solver_settings here
+        mParam.ValidateAndAssignDefaults(default_params);
+    }
+
+    ~EigenGeneralizedEigenvalueSolver() override {}
+
+
+    /**
+     * Solve for all eigenvalues and vectors of the generalized eigenvalue problem.
+     * for self adjoint version:
+     * rA is self adjoint matrix; rB positive-definite matrix.
+     */
+    void Solve(
+            SparseMatrixType& rA,
```

Comment:
It would be better if this method accepts only dense matices. This would avoid a lot of conversions.

Reply:
I agree here, but:

I tried to implement this sub solution as another linear solver, because i think it makes sense to maintain this "solver" property and making use of the factory and so on.
I was not able to create a solver for dense matrices only, also i don't know a smart way preventing users from using it with SparseMatrices then (As they are used to do so with all other solvers).

Maybe it is not even necessary that this dense eigen value solver inherits from the linear_solver.
I glad for any suggestion here.

Reply:
I could think of creating a new class for dense generalized eigenvalue problems:

```dense_generalized_eigenvalue_solver.h```
that implements a  solve function that solves for all eigenvalues of a small dense problem:

``` cpp
void Solve(DenseMatrixType& rA, DenseMatrixType& rB, VectorType& rEigenvalues, DenseMatrixType& rEigenvectors)
```

This class would then not be related to the linear_solvers.h but implement similar functions.
The ```eigen_generalized_eigenvalue_solver``` could then inherit from this new class.
And in the core one could implement e.g.  a ```generalized_jacobi_iteration_eigenvalue_solver``` as it is also done in the Bathe book.

What do you think about that? @pooyan-dadvand  @RiccardoRossi  @philbucher

Reply:
I like the idea
Especially the part of the solve function only accepting Dense matrices makes it (almost :) ) impossible to misuse

Reply:
I think here the misleading is the name of the sparse and dense types. Because the solvers are designed to use different spaces and one can instantiate them with two dense spaces. So the sparse eigen solver would have a dense eigen solver (also optimized for dense) without having to  duplicate the methods. So one solution would be to change those typenames. (Here I would also ping @RiccardoRossi)